src/tableofcontentbar.cpp

```diff
@@ -0,0 +1,100 @@
+    ui->titleLabel->setText(gt("table-of-content"));
+    ui->hideLabel->setFont(QFont("Selawik", 12));
+    ui->hideLabel->setTextFormat(Qt::RichText);
+
+    /* href is needed to make hide clickable, but not used. So Kiwix it is :) */
+    ui->hideLabel->setText("<a href=\"https://kiwix.org/en/\">" + gt("hide") + "</a>");
+    connect(ui->hideLabel, &QLabel::linkActivated, this, [=](){
+        KiwixApp::instance()->getAction(KiwixApp::ToggleTOCAction)->setChecked(false);
+    });
+
+    ui->tree->setRootIsDecorated(false);
+    ui->tree->setItemsExpandable(false);
+    connect(ui->tree, &QTreeWidget::itemClicked, this, &TableOfContentBar::onTreeItemActivated);
+    connect(ui->tree, &QTreeWidget::itemActivated, this, &TableOfContentBar::onTreeItemActivated);
+}
+
+TableOfContentBar::~TableOfContentBar()
+{
+    delete ui;
+}
+
+void TableOfContentBar::onTreeItemActivated(QTreeWidgetItem *item)
+{
+    emit navigationRequested(m_url, item->data(0, AnchorRole).toString());
+}
+
+namespace
+{
+
+QTreeWidgetItem* createChildItem(QTreeWidgetItem* parent, const QJsonObject& headerObj)
+{
+    const auto item = new QTreeWidgetItem(parent);
+    item->setExpanded(true);
+
+    auto numberList = parent->data(0, NumberListRole).toStringList();
+    numberList.append(QString::number(parent->childCount()));
+    item->setData(0, NumberListRole, numberList);
+
+    const auto itemNum = numberList.join(".");
+    const auto display = itemNum + "  " + headerObj["text"].toString();
+    item->setToolTip(0, display);
+    item->setData(0, Qt::DisplayRole, display);
+    item->setData(0, Qt::FontRole, QFont("Selawik", 12));
+    item->setData(0, LevelRole, headerObj["level"].toInt());
+    item->setData(0, AnchorRole, headerObj["anchor"].toString());
+
+    return item;
+}
+
+}
+
+void TableOfContentBar::setupTree(const QJsonObject& headers)
+{
+    const auto headerUrl = headers["url"].toString();
+    const auto webView = KiwixApp::instance()->getTabWidget()->currentWebView();
+    const auto currentUrl = webView->url().url(QUrl::RemoveFragment);
+    if (headerUrl != currentUrl)
+        return;
+    
+    m_url = headerUrl;
+    ui->tree->clear();
+    QJsonArray headerArr = headers["headers"].toArray();
+    QTreeWidgetItem *curParentItem = ui->tree->invisibleRootItem();
+    curParentItem->setData(0, LevelRole, 0);
+    curParentItem->setData(0, NumberListRole, QStringList{});
+
+    for (const auto& header : headerArr)
+    {
+        const auto headerObj = header.toObject();
+        const int headerLevel = headerObj["level"].toInt();
+
+        /* find correct parent level to insert */
+        while (headerLevel <= curParentItem->data(0, LevelRole).toInt())
+        {
+            curParentItem = curParentItem->parent();
+
+            /* Unfortunate that top-level items not parented by invisibleRoot */
+            if (!curParentItem)
+                curParentItem = ui->tree->invisibleRootItem();
+        }
```

Comment:
What are the assumptions about level values in the header data? Can there be gaps like in the following sequence: `{h1, h3, h4, h4, h2, h4}` ? Does your algorithm handle such cases? Otherwise, assuming absence of gaps i.e. all levels on a branch from the root to the leaf are filled, why do you need `LevelRole`?

Reply:
Yes there can be gaps. During parsing, we use their level to determine the correct parent, but how the tree is constructed  as if there are no gaps ( i.e. `{ h1, h3 }` will be repesented in the list as if it is `{ h1, h2 }`). Having perfect headers is an unlikely assumption...

Reply:
Well, I see now that probably under such conditions combining recursion with iteration makes more sense. Without recursion you have to store in the tree items intermediate data like level and the number-list (this one was present in your previous implementation too). Here is a cleaner and more comprehensible scheme:

```c++
QJsonArray takeDeeperEntries(QJsonArray& headerArr, int level)
{
    QJsonArray result;
    while (!headerArr.isEmpty() ) {
        const auto& nextHeader = headerArr.first().toObject();
        if ( nextHeader["level"].toInt() <= level )
           break;
        result.push_back(nextHeader);
        headerArr.pop_front();
    }
    return result;
}

void createSubTree(QTreeWidgetItem* parent, QString parentNo, QJsonArray headerArr)
{
    while (!headerArr.isEmpty()) {
        const auto& childHeader = headerArr.takeAt(0).toObject();
        const int childLevel = childHeader["level"].toInt();
        const QString childNo = parentNo + "." + QString::number(parent->childCount() + 1);
        QTreeWidgetItem* childItem = createChildItem(parent, childNo, childHeader);
        const QJsonArray deeperStuff = takeDeeperEntries(headerArr, childLevel);
        setupSubTree(childItem, childNo, deeperStuff);
    }
}
```

Reply:
Done